### ISSUE TITLE
[Frontend] Linear Type System

### ISSUE THREAD

**Issue by magnus-madsen** [flix/flix#516]
***What:***
Implement a linear type for the functional language.

***Prerequisites:***
The parser (and phases) thereafter have to be extended with a `lin` keyword (or similar).

***Detailed Description:***
We want to introduce a linear (or ownership) type system to track resources such as file handles, input streams, sockets etc. which must not be duplicated and must be released to the operating system when they are no longer used. Unlike Rust, we aim not to track the linearity of every memory allocation (which is not suitable for a functional language with GC), but only for those objects which are actual resources.

For example, we would like to write:

```scala
def readFile(f: lin File): Str = ...
```
where the `lin` modifier on `f` instructs the compiler that `f` must be treated as linear. 

In other words, `f` must be used (and freed) by `readFile`. Moreover, `f` cannot be duplicated.

The subject of linearity and ownership is huge. We should start with something simple that is implementable and can provide some benefit. For example, resource management for an input stream.

***Technical Challenges:***
- We want to keep linearity / borrowing / life times  separate from the type checker.
